exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/CreateFunctionHandler.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.drill.exec.store.sys.store.DataChangeVersion;
+import org.apache.drill.exec.util.JarUtil;
+import org.apache.drill.exec.work.foreman.ForemanSetupException;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.Path;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+public class CreateFunctionHandler extends DefaultSqlHandler {
+
+  private static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(CreateFunctionHandler.class);
+
+  public CreateFunctionHandler(SqlHandlerConfig config) {
+    super(config);
+  }
+
+  /**
+   * Creates UDFs dynamically.
+   *
+   * @return - Single row indicating list of registered UDFs, raise exception otherwise
+   */
+  @Override
+  public PhysicalPlan getPlan(SqlNode sqlNode) throws ForemanSetupException, IOException {
+    if (context.getOption(ExecConstants.DYNAMIC_UDF_SUPPORT_ENABLED).bool_val) {
+      SqlCreateFunction node = unwrap(sqlNode, SqlCreateFunction.class);
+      String jarName = ((SqlCharStringLiteral) node.getJar()).toValue();
+      String sourceName = JarUtil.getSourceName(jarName);
+
+      RemoteFunctionRegistry remoteRegistry = context.getRemoteFunctionRegistry();
+      FileSystem fs = remoteRegistry.getFs();
+      Path tmpDir = new Path(remoteRegistry.getTmpArea(), UUID.randomUUID().toString());
+      File localTmpDir = Files.createTempDir();
+
+      boolean inProgress = false;
+      try {
+        final String action = remoteRegistry.addToJars(jarName, RemoteFunctionRegistry.Action.REGISTRATION);
+        if (!(inProgress = action == null)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Jar with %s name is used. Action: %s", jarName, action));
+        }
+
+        // verify that binary and source exist
+        Path remoteBinary = new Path(remoteRegistry.getStagingArea(), jarName);
+        Path remoteSource = new Path(remoteRegistry.getStagingArea(), sourceName);
+        if (!fs.exists(remoteBinary) || !fs.exists(remoteSource)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Binary [%s] or source [%s] is absent in udf staging area [%s].", jarName, sourceName, remoteRegistry.getStagingArea().toUri().getPath()));
+        }
+
+        // backup binary & source (copy to udf tmp directory)
+        fs.mkdirs(tmpDir);
+        Path tmpBinary = new Path(tmpDir, jarName);
+        Path tmpSource = new Path(tmpDir, sourceName);
+
+        FileUtil.copy(fs, remoteBinary, fs, tmpBinary, false, fs.getConf());
+        FileUtil.copy(fs, remoteSource, fs, tmpSource, false, fs.getConf());
+
+        // copy binary to local fs, we don't need source for validation
+        Path localBinary = new Path(new Path(localTmpDir.toURI()), jarName);
+        fs.copyToLocalFile(tmpBinary, localBinary);
+
+        // validate functions locally
```

Comment:
Clients X changes won't be overwritten, zookeeper versioning will take care about it. See comments below.